pkg/autodiscovery/listeners/cloudfoundry.go

```diff
@@ -0,0 +1,242 @@
+	adIdentifier   cloudfoundry.ADIdentifier
+	containerIPs   map[string]string
+	containerPorts []ContainerPort
+	creationTime   integration.CreationTime
+}
+
+// Make sure CloudFoundryService implements the Service interface
+var _ Service = &CloudFoundryService{}
+
+// NewCloudFoundryListener creates a CloudFoundryListener
+func NewCloudFoundryListener() (ServiceListener, error) {
+	bbsCache, err := cloudfoundry.GetGlobalBBSCache()
+	if err != nil {
+		return nil, err
+	}
+	return &CloudFoundryListener{
+		services:      map[string]Service{},
+		stop:          make(chan bool),
+		refreshTicker: time.NewTicker(10 * time.Second),
+		bbsCache:      bbsCache,
+	}, nil
+}
+
+// Listen periodically refreshes services from global BBS API cache
+func (l *CloudFoundryListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go func() {
+		l.refreshServices(true)
+		for {
+			select {
+			case <-l.stop:
+				l.refreshTicker.Stop()
+				return
+			case <-l.refreshTicker.C:
+				l.refreshServices(false)
+			}
+		}
+	}()
+}
+
+func (l *CloudFoundryListener) refreshServices(firstRun bool) {
+	log.Debug("Refreshing services via CloudFoundryListener")
+	// make sure that we can't have two simultaneous runs of this function
+	l.Lock()
+	defer l.Unlock()
+	allActualLRPs, desiredLRPs := l.bbsCache.GetAllLRPs()
+
+	// if not found and running, add it
+	// at the end, compare what we saw and what is cached and kill what's not there anymore
+	notSeen := make(map[string]interface{})
+	for i := range l.services {
+		notSeen[i] = nil
+	}
+
+	adIdentifiers := l.getAllADIdentifiers(desiredLRPs, allActualLRPs)
+	for _, id := range adIdentifiers {
+		strID := id.String()
+		if _, found := l.services[strID]; found {
+			// delete is no-op when we try to delete a key that doesn't exist
+			// NOTE: this will remove old versions of services on redeploys because ADIdentifier contains ProcessGUID,
+			//       which changes by redeploying
+			delete(notSeen, strID)
+			continue
+		}
+		svc := l.createService(id, firstRun)
+		// if the container is not in RUNNING state, we can't populate ports, thus createService would return nil
+		if svc != nil {
+			l.newService <- svc
+		}
+	}
+
+	for adID := range notSeen {
+		l.delService <- l.services[adID]
+		delete(l.services, adID)
+	}
+}
+
+func (l *CloudFoundryListener) createService(adID cloudfoundry.ADIdentifier, firstRun bool) *CloudFoundryService {
+	crTime := integration.After
+	if firstRun {
+		crTime = integration.Before
+	}
+
+	var svc *CloudFoundryService
+	aLRP := adID.GetActualLRP()
+	if aLRP == nil {
+		// non-container service
+		// NOTE: non-container services intentionally have no IPs or ports, everything is supposed to be configured
+		// through the "variables" section in the AD configuration
+		svc = &CloudFoundryService{
+			adIdentifier:   adID,
+			containerIPs:   map[string]string{},
+			containerPorts: []ContainerPort{},
```

Comment:
this means `%%host%%` won't be supported in this case

Reply:
Yeah, that is expected. The only templating variables for non-container services will be the ones from `variables` defined and rendered in the provider. I'll make sure this is documented properly.